### PR TITLE
eos : Add resource module eos_acl_interfaces

### PR DESCRIPTION
**Pull request by GomathiselviS**
##### SUMMARY
This adds the resource module eos_acl_interfaces

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
eos_acl_interfaces.py

### CODE REVIEW

File: lib/ansible/module_utils/network/common/facts/facts.py [hunk 1]
```diff
@@ -23,7 +23,6 @@ def __init__(self, module):
         self._connection = None
         if module.params.get('state') not in ['rendered', 'parsed']:
             self._connection = get_resource_connection(module)
-
```
Comment (pabelanger) [inline lib/ansible/module_utils/network/common/facts/facts.py]: this looks unrelated?

File: lib/ansible/module_utils/network/eos/eos.py [hunk 2]
```diff
@@ -129,6 +129,8 @@ def run_commands(self, commands, check_rc=True):
         """Run list of commands on remote device and return results
         """
         connection = self._get_connection()
+        import q
+        q(commands)
```
Comment (pabelanger) [inline lib/ansible/module_utils/network/eos/eos.py]: this can be deleted

File: lib/ansible/plugins/filter/mathstuff.py [hunk 3]
```diff
@@ -111,6 +111,8 @@ def symmetric_difference(environment, a, b):
     else:
         isect = intersect(environment, a, b)
         c = [x for x in union(environment, a, b) if x not in isect]
+    import q
+    q(a, b, c)
```
Comment (pabelanger) [inline lib/ansible/plugins/filter/mathstuff.py]: same

File: test/integration/targets/eos_acl_interfaces/meta/main.yaml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+dependencies:
+  - prepare_ios_tests
```
Comment (pabelanger) [inline test/integration/targets/eos_acl_interfaces/meta/main.yaml]: This should be prepare_eos, not ios.

Comment (GomathiselviS) [reply test/integration/targets/eos_acl_interfaces/meta/main.yaml]: Done

File: test/integration/targets/eos_acl_interfaces/defaults/main.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+---
+testcase: "[^_].*"
```
Comment (pabelanger) [inline test/integration/targets/eos_acl_interfaces/defaults/main.yaml]: testcase: ""

File: lib/ansible/module_utils/network/eos/argspec/acl_interfaces/acl_interfaces.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            'options': {
+                'access_groups': {
+                    'elements': 'dict',
+                    'options': {
+                        'acls': {
+                            'elements': 'dict',
+                            'options': {
+                                'direction': {
+                                    'choices': ['in', 'out'],
+                                    'type': 'str'
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/eos/argspec/acl_interfaces/acl_interfaces.py:51]: Shouldn't this be required too?

File: lib/ansible/module_utils/network/eos/argspec/acl_interfaces/acl_interfaces.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                                'name': {
+                                    'required': True,
+                                    'type': 'str'
+                                }
+                            },
+                            'type': 'list'
+                        },
+                        'afi': {
+                            'choices': ['ipv4', 'ipv6'],
+                            'type': 'str'
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/eos/argspec/acl_interfaces/acl_interfaces.py:63]: Same as above. I believe this can also be required: True.

File: lib/ansible/module_utils/network/eos/facts/acl_interfaces/acl_interfaces.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        objs = []
+        for resource in resources:
+            if resource:
+                obj = self.render_config(self.generated_spec, resource)
+                if obj:
+                    objs.append(obj)
+
+        ansible_facts['ansible_network_resources'].pop('acl_interfaces', None)
+        facts = {}
+        if objs:
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/eos/facts/acl_interfaces/acl_interfaces.py:72]: Will this check prevent any interfaces from being rendered if no acls are configured on them?

File: lib/ansible/modules/network/eos/eos_acl_interfaces.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+#
+# eos#sh running-config | include interface|access-group
+# interface Ethernet1
+# interface Ethernet2
+#  ip access-group acl01 in
+#  ipv6 access-group acl03 out
+# interface Ethernet3
+#  ip access-group acl01 out
+
+- name: "Delete module attributes of given access-groups"
```
Comment (NilashishC) [inline lib/ansible/modules/network/eos/eos_acl_interfaces.py:259]: Could you also add an example that shows how to delete all acls belonging to a particular AFI for an interface?

File: lib/ansible/modules/network/eos/eos_acl_interfaces.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+  returned: when changed
+  type: list
+  sample: >
+    The configuration returned will always be in the same format
+     of the parameters above.
+commands:
+  description: The set of commands pushed to the remote device.
+  returned: always
+  type: list
+  sample: ['command 1', 'command 2', 'command 3']
```
Comment (NilashishC) [inline lib/ansible/modules/network/eos/eos_acl_interfaces.py]: Please update the sample commands.

File: lib/ansible/modules/network/eos/eos_acl_interfaces.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+from ansible.module_utils.network.eos.argspec.acl_interfaces.acl_interfaces import Acl_interfacesArgs
+from ansible.module_utils.network.eos.config.acl_interfaces.acl_interfaces import Acl_interfaces
+
+
+def main():
+    """
+    Main entry point for module execution
+
+    :returns: the result form module invocation
+    """
```
Comment (NilashishC) [inline lib/ansible/modules/network/eos/eos_acl_interfaces.py:406]: The required_if checks are absent.

File: lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+
+        :param want: the desired configuration as a dictionary
+        :param have: the current configuration as a dictionary
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        if self.state in ('merged', 'replaced', 'overridden') and not want:
+            self._module.fail_json(msg='value of config parameter must not be empty for state {0}'.format(self.state))
```
Comment (NilashishC) [inline lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py:124]: Please add an empty_config.yaml test file to assert this for all the states.

File: lib/ansible/module_utils/network/eos/facts/acl_interfaces/acl_interfaces.py (hunk 8 above)
Comment (GomathiselviS) [reply lib/ansible/module_utils/network/eos/facts/acl_interfaces/acl_interfaces.py:72]: No, this will not prevent the interfaces from being rendered, when there are no acls attached to them. I hope that is the expectation.

**Review by NilashishC — APPROVED** [review]

File: test/integration/targets/eos_acl_interfaces/tests/cli/deleted.yaml [hunk 13]
```diff
@@ -0,0 +1,57 @@
+---
+- debug:
+    msg: "Start eos_acl_interfaces deleted integration tests ansible_connection={{ ansible_connection }}"
+
+- include_tasks: _populate.yaml
+
+- block:
+
+  - name: Delete attributes of given acl_interfaces of an afi.
```
Comment (justjais) [inline test/integration/targets/eos_acl_interfaces/tests/cli/deleted.yaml:9]: can u include idempotency check for delete scenarios as well.

File: lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        """ Select the appropriate function based on the state provided
+
+        :param want: the desired configuration as a dictionary
+        :param have: the current configuration as a dictionary
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        if self.state in ('merged', 'replaced', 'overridden') and not want:
```
Comment (justjais) [inline lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py:123]: Please add `rendered` state as well

**Review by justjais — COMMENTED** [review]
minor cosmetic fix

File: lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py (hunk 14 above)
Comment (justjais) [reply lib/ansible/module_utils/network/eos/config/acl_interfaces/acl_interfaces.py:123]: ah didn't see it merged already!
